### PR TITLE
cilium: Remove attached bpf_xdp upon "cilium cleanup"

### LINKED ISSUE #19586: Remove attached bpf_xdp upon "cilium cleanup"
`cilium cleanup` doesn't remove a previously attached XDP progs by Cilium. This will lead to packet drops, as the cmd removes the prog maps leading to the XDP prog missing tail calls.

The removal of the XDP prog should happen before the maps removal.

Similar to https://github.com/cilium/cilium/issues/10067.

### PR DESCRIPTION
**Pull request by zhanghe9702**
cilium cleanup doesn't remove a previously attached XDP progs by Cilium,
this commit is an complementary.

Fixes: #19586

Signed-off-by: zhang he <zhanghe9702@163.com>

### CODE REVIEW

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Hi, sorry for the late response and thanks for your contribution!

Could you use `vishvananda/netlink` and `cilium/ebpf` instead of `ip` and `bpftool`? We are now basically removing those two from our code base.

https://github.com/cilium/cilium/pull/19159

**Comment by zhanghe9702** [comment]
that's no problem! i will update commit

**Comment by zhanghe9702** [comment]
@YutaroHayakawa , as far as I know, BPF_PROG_GET_FD_BY_ID/BPF_OBJ_GET_INFO_BY_FD equivalence cilium/ebpf call is

-  https://github.com/cilium/ebpf/blob/358b2b4d7b9f2fc937daadcacfd07d58be448760/internal/sys/types.go#L887
- https://github.com/cilium/ebpf/blob/358b2b4d7b9f2fc937daadcacfd07d58be448760/internal/sys/types.go#L834
but it is  go internal package function, so cilium/cilium can't use it directly! :-)
@brb suggest me using `bpftool show prog id` to retrive cilium special XDP infomation when clean XDP, what's your suggestion about it?

**Comment by YutaroHayakawa** [comment]
Ah, alright. In that case, please use `bpftool` for now. We appreciate if you add support for that :)

**Comment by zhanghe9702** [comment]
i could add an `bpftool show prog id` equivalent func in link/program.go, do you have a better recommendation?

**Comment by brb** [comment]
@zhanghe9702 Thanks for the PR! 

> but it is go internal package function, so cilium/cilium can't use it directly! :-)

Maybe you could ping https://github.com/cilium/ebpf maintainers about exporting the functions (e.g., @ti-mo)? It would be good to not a new code which depends on tools from which we are trying to migrate off.

File: cilium/cmd/cleanup.go [hunk 1]
```diff
@@ -175,6 +186,12 @@ func (c ciliumCleanup) whatWillBeRemoved() []string {
 		}
 		toBeRemoved = append(toBeRemoved, section)
 	}
+	if len(c.xdpLinks) > 0 {
+		section := "xdps\n"
```
Comment (ti-mo) [inline cilium/cmd/cleanup.go]:
```suggestion
		section := "xdp programs\n"
```

File: bpf/bpf_xdp.c [hunk 2]
```diff
@@ -274,7 +274,7 @@ static __always_inline int check_filters(struct __ctx_buff *ctx)
 }
 
 __section("from-netdev")
-int bpf_xdp_entry(struct __ctx_buff *ctx)
+int cilium_bpf_xdp_entry(struct __ctx_buff *ctx)
```
Comment (ti-mo) [inline bpf/bpf_xdp.c]: Please leave this untouched for now, we should make this kind of change throughout the codebase. Changing one of them here will make it even more inconsistent.

File: cilium/cmd/cleanup.go [hunk 3]
```diff
@@ -511,3 +531,46 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
...
+
+	args := []string{"prog", "show", "id", fmt.Sprintf("%d", progId)}
+	output, err := exec.Command(prog, args...).CombinedOutput()
+	if err != nil {
+		return false, err
+	}
+
+	s := strings.Fields(string(output))
+	for i := range s {
+		if strings.HasPrefix(s[i], ciliumLinkPrefix) {
```
Comment (ti-mo) [inline cilium/cmd/cleanup.go]: Please just match on `bpf_xdp_entry` here so this pops up when I grep it later. Finding the resources Cilium attaches on the system is already tricky enough, let's not complicate matters by abusing `ciliumLinkPrefix`.

**Review by ti-mo — CHANGES_REQUESTED** [review]
Please `s/Xdp/XDP/` everywhere. (except for `netlink.LinkSetXdpFd`, of course)

File: cilium/cmd/cleanup.go [hunk 4]
```diff
@@ -511,3 +533,48 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
...
 }
+
+func removeXDPs(links []netlink.Link) error {
+	for _, link := range links {
+		err := netlink.LinkSetXdpFd(link, -1)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("removed cilium xdp of %s\n", link.Attrs().Name)
+
```
Comment (YutaroHayakawa) [inline cilium/cmd/cleanup.go]: Nit: Don't need this line break.

**Review by YutaroHayakawa — APPROVED** [review]
LGTM with Nit

File: cilium/cmd/cleanup.go [hunk 5]
```diff
@@ -161,8 +164,17 @@ func newCiliumCleanup(bpfOnly bool) ciliumCleanup {
...
 		}
 	}
+	xdpLinks := []netlink.Link{}
 
-	return ciliumCleanup{routes, ciliumLinks, tcFilters, netNSs, bpfOnly}
+	for _, link := range links {
+		if ok, err := isLinkAttachedCiliumXDP(link); ok && err == nil {
+			xdpLinks = append(xdpLinks, link)
+		} else if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
```
Comment (twpayne) [inline cilium/cmd/cleanup.go:173]: This should be logged, rather than written to os.Stderr.

File: cilium/cmd/cleanup.go [hunk 6]
```diff
@@ -511,3 +533,47 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
...
 	return nil
 }
+
+func removeXDPs(links []netlink.Link) error {
+	for _, link := range links {
+		err := netlink.LinkSetXdpFd(link, -1)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("removed cilium xdp of %s\n", link.Attrs().Name)
```
Comment (twpayne) [inline cilium/cmd/cleanup.go:544]: This should be logged, rather than written to os.Stdout.

**Review by twpayne — APPROVED** [review]

**Review by ti-mo — APPROVED** [review]

File: cilium/cmd/cleanup.go [hunk 7]
```diff
@@ -511,3 +533,47 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
...
+	if err != nil {
+		return false, err
+	}
+
+	info, err := xdp.Info()
+	if err != nil {
+		return false, err
+	}
+
+	if strings.Contains(info.Name, "bpf_xdp_entry") {
```
Comment (brb) [inline cilium/cmd/cleanup.go]: I am a bit paranoid that others might use the same entry point name. Mind renaming?

**Review by brb — CHANGES_REQUESTED** [review]

File: cilium/cmd/cleanup.go (hunk 7 above)
Comment (brb) [reply cilium/cmd/cleanup.go]: Oh, I see https://github.com/cilium/cilium/pull/19735#discussion_r879625788. @ti-mo mind elaborating what's your concern regarding changing the function name? Also, cc @borkmann just in case I'm missing something.

Comment (ti-mo) [reply cilium/cmd/cleanup.go]: We currently don't prefix any bpf entrypoint function names with `cilium_`, so if we want to change it, we should change it across the whole codebase so it's consistent. A change like this should be discussed and planned, as we might for example, want to go with `cil_` (and remove the redundant `bpf_` prefix -- yeah, it's a bpf program alright!) to try to stay under the bpf_obj_name length limit as much as possible. See a similar discussion at https://github.com/cilium/cilium/issues/10067#issuecomment-1024322477. (cc @YutaroHayakawa @pchaigno)

Also, I started relying on the entrypoint name to classify the program type in my Go ELF loader patch: https://github.com/cilium/cilium/pull/19159/commits/f5db63ea6de9a51c31ca987e73630a8845fcc9ad#diff-d6f11688bc0f4227c8d78948312a1ca6e290c1c63dcf8c0db12a45d035fb9d5dR139-R140. We should switch to libbpf section name prefixes here to classify them automatically, but it's just an example of something that might break when changing this.

tl;dr: let's discuss this first, please. This would be a good topic to revisit when we move off of the iproute2 cli, since e.g. ELF obj names and function names are used as the qdisc name, and those currently don't have any 'cilium' identifiers.

Comment (brb) [reply cilium/cmd/cleanup.go]: Let's discuss this with the wider team at the sig-datapath meeting on Thu.

File: cilium/cmd/cleanup.go [hunk 8]
```diff
@@ -511,3 +533,47 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
...
+		err := netlink.LinkSetXdpFd(link, -1)
+		if err != nil {
+			return err
+		}
+		log.Infof("removed cilium xdp of %s", link.Attrs().Name)
+	}
+	return nil
+}
+
+func isLinkAttachedCiliumXDP(link netlink.Link) (bool, error) {
```
Comment (pchaigno) [inline cilium/cmd/cleanup.go]: Maybe `isCiliumXDPAttachedToLink`?

File: cilium/cmd/cleanup.go [hunk 9]
```diff
@@ -511,3 +533,47 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
...
 	return nil
 }
+
+func removeXDPs(links []netlink.Link) error {
+	for _, link := range links {
+		err := netlink.LinkSetXdpFd(link, -1)
+		if err != nil {
+			return err
+		}
+		log.Infof("removed cilium xdp of %s", link.Attrs().Name)
```
Comment (pchaigno) [inline cilium/cmd/cleanup.go]: I don't see other functions using `log.Info` but `fmt.Printf` instead. Which should we be using?

File: cilium/cmd/cleanup.go (hunk 7 above)
Comment (pchaigno) [inline cilium/cmd/cleanup.go]: How can we ensure that we won't forget to update this if/when we update the function's name in `bpf_xdp.c`? AFAIK, we don't have any tests covering this. At the very least, we should add a comment in `bpf_xdp.c`.

**Review by pchaigno — COMMENTED** [review]
A couple minor comments below, but looks good to me otherwise. Thanks for tackling this! :rocket:

File: cilium/cmd/cleanup.go (hunk 7 above)
Comment (brb) [reply cilium/cmd/cleanup.go]: I've added to this week's sig-datapath agenda to discuss the renaming https://docs.google.com/document/d/1IZZ-FTM_U3zN06-M_mDOgep_DcDM0Oe3YOFVM1BTIt8/edit#heading=h.f7lmu4rc837r.

Comment (brb) [reply cilium/cmd/cleanup.go]: @zhanghe9702 We have discussed the issue, and decided to prefix the entry function with `cli_`. Perhaps you could rename it to `cli_xdp_entry` which should not exceed the 15 char limit.

Comment (pchaigno) [reply cilium/cmd/cleanup.go]: @brb I think you mean `cil_` and `cil_xdp_entry`.

Comment (zhanghe9702) [reply cilium/cmd/cleanup.go]: got it! i will update code when i  back to my city at MONDAY :-)

**Review by brb — APPROVED** [review]
Thanks, LGTM.

File: cilium/cmd/cleanup.go (hunk 9 above)
Comment (pchaigno) [reply cilium/cmd/cleanup.go]: Let's use `fmt.Printf` here. I'm not sure `log.Info` would print anything given it's a CLI.

Same for the `log.WithError(err).Errorf` above.

File: cilium/cmd/cleanup.go (hunk 7 above)
Comment (pchaigno) [reply cilium/cmd/cleanup.go]: The comment is still needed IMO.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Logging still needs to be fixed. See previous review.

**Review by pchaigno — APPROVED** [review]

**Comment by tklauser** [comment]
/test-1.22-5.4
